02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md

```diff
@@ -0,0 +1,550 @@
+# State of PureScript Documentation: 2019
+
+This document is not meant to convince you to use PureScript, we assume that readers are already interested in it or using it.
+
+It is well-known amongst the PureScript community that its documentation is lacking in some critical areas. As a result, @chexxor started the [PureScript documentation efforts in 2019](https://discourse.purescript.org/t/purescript-documentation-efforts-in-2019/524) Discourse thread to answer one question:
+> How can we improve the PureScript documentation in 2019?
+
+(Purpose) The following document tries to do the following things:
+1. Provide a researched narrative of PureScript's documentation development and history.
+2. Define the types of audiences of PureScript documentation and their expectations.
+    - New learners (regardless of language background)
+    - PureScript documentation writers
+    - Core contributors
+3. Summarize the current opinions of PureScript's documentation audiences.
+4. Explore strategies for improving PureScript's documentation for its audiences.
+
+<!-- https://imthenachoman.github.io/nGitHubTOC/ -->
+
+- [What is "Good" Documentation Anyway?](#what-is-good-documentation-anyway)
+  - [The Types of Documentation](#the-types-of-documentation)
+  - [The Documentation's Intended Audience](#the-documentations-intended-audience)
+  - [Maintaining Documentation's Accuracy](#maintaining-documentations-accuracy)
+    - [The "Size" of a Change](#the-size-of-a-change)
+    - [The "Frequency" of a Change](#the-frequency-of-a-change)
+    - [How to Make Maintenance Easier](#how-to-make-maintenance-easier)
+  - [Criteria for "Good" Documentation](#criteria-for-good-documentation)
+  - [Evaluating PureScript's Documentation](#evaluating-purescripts-documentation)
```

Comment:
I think this TOC and the document organization should reflect the order of the items in the "Purpose" statement. When I drafted that "Purpose" the order of its items seemed to naturally form a natural narrative (I haven't considered how this narrative compares with the GTD process or other narratives we have developed during this project):

1. What is current PureScript documentation situation?
2. Who is reading the documentation?
3. What do the readers think about it?
4. How can we improve the documentation for these readers, given their opinions and our understanding of what other documentation authors think documentation should look like?

Anyways, if we adjust the order of this document to follow the narrative in the Purpose, I think this should come first.

Reply:
To help visualize that re-arrangement, here's one way we could reorder the sections in this document.

Note how the "Improvement Strategies" is empty here. I think most of the content I previously thought was strategy development is actually part of understanding a potential target state. Developing strategies for moving towards that target, then, requires further work.

```
- 1. Current PureScript Documentation
  - [Evaluating PureScript's Documentation](#evaluating-purescripts-documentation)
  - [Why is PureScript's Documentation Lacking and How Do We Improve It?](#why-is-purescripts-documentation-lacking-and-how-do-we-improve-it)
- 2, 3. Audiences and Opinions
  - [New Learners: What is the Best Way to Learn PureScript?](#new-learners-what-is-the-best-way-to-learn-purescript)
  - [Core Contributors: ???](#core-contributors-)
  - [PureScript Documentation Writers: What is the Best Way to Write Documentation in this Context?](#purescript-documentation-writers-what-is-the-best-way-to-write-documentation-in-this-context)
- 4. Documentation Improvement
  - [What is "Good" Documentation Anyway?](#what-is-good-documentation-anyway)
    - [The Types of Documentation](#the-types-of-documentation)
    - [The Documentation's Intended Audience](#the-documentations-intended-audience)
    - [Maintaining Documentation's Accuracy](#maintaining-documentations-accuracy)
      - [The "Size" of a Change](#the-size-of-a-change)
      - [The "Frequency" of a Change](#the-frequency-of-a-change)
      - [How to Make Maintenance Easier](#how-to-make-maintenance-easier)
    - [Criteria for "Good" Documentation](#criteria-for-good-documentation)
  - Improvement Strategies (To Do)
```

Reply:
For context, [this was the original approach to this file](https://github.com/chexxor/purescript-documentation-discussion/pull/50#issuecomment-468945658)

> I think this TOC and the document organization should reflect the order of the items in the "Purpose" statement. 

I agree with this. However, I disagree slightly on what your rendering is for the document's purpose / your organization of the document's structure. I'll explain more of my points below.

First thought: 
In our current work, the `why are PS docs lacking` section seems to be answering primarily one question, "Why can't we improve the library docs that get published on Pursuit via PRs / improve other websites related to documentation?" (answer: the people who have write-access are spread-thin, so such PRs get forgotten/delayed. Oh, and the bower OOM was also an issue but was fortunately fixed recently). 

However, of the 5 types of documentation, the Pursuit docs issue (aside from the website stuff) is only one type: the reference type. 
The other 2 explanations in that section (the support system not building towards persistent structured docs; FP concepts are hard to define) refer to other types of docs: how-to guides and explanations. These explanations should be relocated into their own section.

Second thought:
Logically, it's hard to say "X is bad" unless we have criteria by which to judge it. Hence, the 'What is Good Documentation anyway" section provides that criteria. However, that section is long and I doubt people will read through the entire document because of it. In your current approach, it seems you're trying to account for that by placing it at the bottom of the file. But then the 'Evaluating PS docs' appears first without any context as to what those doc type tags even mean.
I think it would be better to summarize that section (similar to what we did with 'why docs are lacking' section) and refer back to the actual details in their own file (like what I did in #61). The summary would simply be the 5 types of docs table because it is
- short, so that it doesn't stop people from reading the rest of the file
- provides a clear criteria

The full details of the 'What is Good documentation anyway' section is really only interesting/applicable to people who write the docs.

Thus, I'd propose a structure like this:

1. Current PureScript Documentation
    - The 5 Types of Documentation Table
        - explains what 'documentation' can mean (5 types) and provides necessary criteria (what makes a type good)
        - links to separate file that includes full content of the rest of that section
    - Describing PureScript's Documentation 
        - tags 1 or more doc types to each documentation
        - We could point out here the types of docs that are lacking (how-to guides, explanations, etc.)
    - Cover the difficulty of defining FP's best practices
    - Cover the lack of persistent structured docs problem (Slack and Discourse issue)
    - Cover "core contributors are spread thin" issue
        - shared explanation between both
        - cover 'why Library docs on Pursuit are lacking' aspect
        - cover 'why website-related things aren't getting updated quickly' aspect
2. Addressing Specific Audiences
    - Core Contributors - ask them for feedback on what support they want/need and allow them to answer that publicly (rather than privately)
    - New Learners section - this remains unchanged: provide our suggestions for how to learn in this context
    - Documentation Writers - just link to the file that (in this situation) now stores the full 'What is Good documentation anyway' section.
3. Documentation Improvement (see [my work in #61 for context](https://github.com/chexxor/purescript-documentation-discussion/blob/664347273e3e640931ce8ee1c6559b79f5388b36/02-Context-or-Narrative/State-of-PS-Docs/State-of-PureScript-Documentation-2019.md#everyone-how-should-we-improve-this-situation))
    - Define the purpose we're trying to achieve
    - Define the principles we'll need to abide by
    - Ask everyone for their ideas for improvement
    - Suggest our ideas for ways to improve situation

The final state of this rendering provides these benefits:
- The document is short, concise, and to the point, but allows those who are curious to read a more in-depth explanation of the parts that interest them.
- Before asking anyone to just list off their ideas for how to improve the docs (Brainstorming), people are told what the principles are by which they must abide to do this well and encouraged to think about outcomes that we then figure out how to do (Principles & Outcomes)

Putting this more into the context of our methodology here, Part 1 is the Context/Narrative. Part 2 is a mixture of various goals because we're trying to address multiple audiences at once. Part 3 is the Purpose-Outcomes part of our methodology.

Ok, I'll stop now! Sorry for the long comment! I hope this is constructive criticism and not annoying.

> Note how the "Improvement Strategies" is empty here. I think most of the content I previously thought was strategy development is actually part of understanding a potential target state. Developing strategies for moving towards that target, then, requires further work.

Does this mean you were thinking about actions we could do (Brainstorm) rather than a state we should try to achieve (Outcomes)?

Reply:
More context, but this was the approach I took in #61 as a visual:
![PS Doc Context](https://user-images.githubusercontent.com/8413037/55371834-7c6ee680-54b4-11e9-9388-96b43ba7d71f.png)

Reply:
> - 2, 3. Audiences and Opinions

I don't see an "Opinions" part in your proposal for the document structure. Do you intend to incorporate the audience's reported opinion's/viewpoints in the "3. Documentation Improvement" section of your proposal? (I'm not exactly sure what "opinions" here refer to, now, so I'll post an update after reviewing that.)

> In our current work, the why are PS docs lacking section seems to be answering primarily one question...
> ...
> However, of the 5 types of documentation, the Pursuit docs issue (aside from the website stuff) is only one type: the reference type.

I also saw some "holes" in this document. I was assuming this was still a first draft, relatively, and we'd come back later after deciding the structure and conclusion of this document (to set us up for Brainstorming ways to achieve the desired Outcome?) to use our Sources to fill in the missing pieces.

> Does this mean you were thinking about actions we could do (Brainstorm) rather than a state we should try to achieve (Outcomes)?

Yeah, I was thinking that. It seems we can keep the scope of this document appropriate by focusing on this document building up towards an "Outcomes" section. I see you are qualifying the Outcome by surfacing the Purposes and Principles immediately before it in your section 3 -- I think this is a good idea and by keeping those in close proximity should help ensure the Outcomes stay in-line with them.

Finally I'd like to note one concern of mine - for the "What is Good documentation anyway" topic we have only two sources. While I really like these two sources because they seem to be written by our peers in this kind of endeavor and they seem to be very well-reasoned, my nature is to reserve my confidence until I have several well-reasoned sources. What do you think about this concern? To keep this moving forward, I propose that I would be quite satisfied if we appropriately qualify the places in this document which lean heavily on those two "What is Good documentation" sources, perhaps as "one person's well-reasoned examination" or something like that.

Reply:
> I don't see an "Opinions" part in your proposal for the document structure. Do you intend to incorporate the audience's reported opinion's/viewpoints in the "3. Documentation Improvement" section of your proposal? (I'm not exactly sure what "opinions" here refer to, now, so I'll post an update after reviewing that.)

That was just due to copying-and-pasting your stuff and modifying it. I don't think I would have included it otherwise as I didn't know what 'opinions' referred to either when I initially read through yours.

> I also saw some "holes" in this document. I was assuming this was still a first draft, relatively, and we'd come back later after deciding the structure and conclusion of this document (to set us up for Brainstorming ways to achieve the desired Outcome?) to use our Sources to fill in the missing pieces.

What 'holes' did you see? Doc types that aren't covered in our research? Or something else?

> Yeah, I was thinking that. 

Ah, ok.

> It seems we can keep the scope of this document appropriate by focusing on this document building up towards an "Outcomes" section. 

Sounds good to me. Assuming readers get to that section, I think they'll be in a better mindset for thinking about what those outcomes might be.

> I see you are qualifying the Outcome by surfacing the Purposes and Principles immediately before it in your section 3 -- I think this is a good idea and by keeping those in close proximity should help ensure the Outcomes stay in-line with them.

Great, let's do that. Now, those were the principles I saw arise when writing this document. Did you come up with any others that might also be relevant?

> Finally I'd like to note one concern of mine - for the "What is Good documentation anyway" topic we have only two sources. While I really like these two sources because they seem to be written by our peers in this kind of endeavor and they seem to be very well-reasoned, my nature is to reserve my confidence until I have several well-reasoned sources. What do you think about this concern? To keep this moving forward, I propose that I would be quite satisfied if we appropriately qualify the places in this document which lean heavily on those two "What is Good documentation" sources, perhaps as "one person's well-reasoned examination" or something like that.

That's a really good point! On one hand, I'd ask how many more sources you'd need to feel comfortable with this? I'm really not sure what else could be said in this regard, but that's only because I can't imagine that, not because there isn't more to say. On another hand, I agree with you. Why should we trust these sources? We're really only trusting them because there's no other sources we could find in this regard.
Regardless, I'm completely fine with adding an opening paragraph that makes note of this issue/concern.

Reply:
> Do you intend to incorporate the audience's reported opinion's/viewpoints in the "3. Documentation Improvement" section of your proposal?

Ah, now I see what you meant by this comment. In reality, there never was a 3 in my proposal. 3 was really 4 (Documentation improvement by exploring possible outcomes to strive towards)

Reply:
> What 'holes' did you see?

These, for example: ```[further explained here]()``` Also, because this document is based largely on a first draft you made when becoming stuck, I am guessing there must be something missing. I **feel** like a lot of the Sources aren't reflected here - I'll need to review that part of this project to verify that, and it's equally likely I'm wrong about that.

Reply:
Oh, those were supposed to refer to their corresponding sections but I didn't want to hard-code the anchor links yet until we we sure that the headers for a section were locked.

For example, the paragraph in [this section](https://github.com/chexxor/purescript-documentation-discussion/blob/2f4ae26e27d6f518671f2608e3896f689ae359c8/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#why-is-purescripts-documentation-lacking-and-how-do-we-improve-it) that mentions PS not trying to become the next mainstream language would refer to [this section](https://github.com/chexxor/purescript-documentation-discussion/blob/2f4ae26e27d6f518671f2608e3896f689ae359c8/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#purescript-is-not-currently-trying-to-be-the-next-mainstream-language) where it's explained in more detail.

> Also, because this document is based largely on a first draft you made when becoming stuck, I am guessing there must be something missing. I feel like a lot of the Sources aren't reflected here - I'll need to review that part of this project to verify that, and it's equally likely I'm wrong about that.

I think it was more of me not being sure how we should structure the document because I was going through my own information overload moment.
That being said, there are a few sources (e.g. sources related to Bower and the 'how to avoid incoherence in the ecosystem' source) that might not be included in here as much.

Reply:
I think the "Audience's Opinions" thing was introduced by me at the start of this thread.

> 3. What do the readers think about it?

Trying to recall now, I think I introduced that because it seems important to mention the opinions of the current documentation state. This document currently starts with "It is well-known amongst the PureScript community that its documentation is lacking in some critical areas." I think we remove this presumption and instead prove it by referencing the opinions we gathered in our Sources. A natural place to do that would be following introducing those audiences, and following the overview of the current documentation state (e.g. Pursuit exists, GH/purescript/documentation exists, blog posts exists, some books exist).

Reply:
Ohh!!!! Right, you've brought that up before when you said something about "even if 99% of people think X is true, it's still their perception that X is true."

> I think we remove this presumption and instead prove it by referencing the opinions we gathered in our Sources. 

Sounds good to me.

> A natural place to do that would be following introducing those audiences, and following the overview of the current documentation state (e.g. Pursuit exists, GH/purescript/documentation exists, blog posts exists, some books exist).

Umm... I'm not understanding you here.

Reply:
That last pull quote is just my trying to justify the reason for placing it where it is - number 3 in my first comment in this thread. :-)

Reply:
Ah! ok.

I think I'd just replace this
> It is well-known amongst the PureScript community that its documentation is lacking in some critical areas.

with this
> Numerous parties within the PureScript community (e.g. [core contributors](), [users](), [new learners]()) believe that the documentation is lacking in some critical areas.

or even this if we have multiple links for each one
> Numerous parties within the PureScript community (e.g. core contributors: [1](), [2](), [3](); users: [1](), [2](); and new learners: [1](), [2]()) believe that the documentation is lacking in some critical areas.

And the above links would point to specific sources where they state that.

Reply:
Oh! Great resolution!

Also, about the empty references you responded to a few comments above. Gotcha 👌.

Reply:
So tasks we've identified here:
- [ ] convert the 'it is well-known amongst the PS community' presumption into a proven statement using the above rendering idea. (We'll need to determine which links to include here)
- [ ] warn about how we're only using 2 sources to define the 'what is good docs' section. Alex, can you do that since I'm not sure how we'd want to word that and you've likely thought more about it than I have?

Where are we at in terms of the document's proposed structure?

Reply:
Working on those two tasks here:
https://github.com/chexxor/purescript-documentation-discussion/pull/63
https://github.com/chexxor/purescript-documentation-discussion/pull/64

Reply:
Also, that document structure should work well. I think we can turn it into a pretty readable narrative.

Reply:
Do you want me to submit a PR to re-organize it to match your proposed way?

Reply:
> Do you want me to submit a PR to re-organize it to match your proposed way?

Sure!

Reply:
Resolved by https://github.com/chexxor/purescript-documentation-discussion/pull/66